Completely rewrite the pull request review guidelines (#558)
space-wizards/docs: src/en/wizden-staff/maintainer/review-procedure.md

Pull requests targeting the `staging` or `stable` branch are subject to the [Hotfix Procedure](hotfix-procedure.md).

Any portion of this procedure may be waived and/or modified with written permission (_via Discord or Github_) from a Lead Maintainer, Project Manager or Wizard.

```admonish note
Note that these procedures are ultimately flexible, however Maintainers that deviate from this policy are expected to provide concrete and valid reasoning when doing so.
Getting one or more other maintainer(s) to agree with your deviation is advisable.

Note that all pull requests that are scheduled for release to `stable` are reviewed during _Stable Review_ in maintainer meetings.
```

## Reviewing Content Pull Requests

### Reviewing the PR
All pull requests must adhere to the [code conventions](../../general-development/codebase-info/conventions.md).

All parts of the Pull Request Template must be filled out by the author.

If the author has obviously not read the conventions or PR guidelines, or the pull request template has not been filled out to a satisfactory level, close the PR.

> All PRs must be triaged, as per the [triage procedure](triage-procedure.md) guidelines.

Reviewer: ```suggestion
All PRs must be triaged according to the [triage procedure](triage-procedure.md) guidelines.
```

When leaving comments or requesting changes:
- Use **constructive** language and avoid being overly negative.
- Try not to be _overly_ vulgar or swear in a derogatory way.
- Avoid comments that personally criticize the author. 
  - It's okay to critique code and decisions, but you shouldn't insult their character.
- For people who may not be familiar with a system, try to be more thorough in your explanations.
- When possible, direct people to either the [developer wiki](https://docs.spacestation14.com/index.html), `#howdoicode` discord channel, or alternate source for information.
- Make an effort to be available after reviewing a PR. 
  - You don't need to be constantly online, but if the author has questions, they should be able to get a response in a reasonable amount of time.
- Be formal when closing PRs. 
  - Discourage others from engaging in rude behavior and try not to upset the PR author. People can often be emotional after their PR is closed.
- For relatively minor changes, opt to simply [complete them yourself](https://cli.github.com/manual/gh_pr_checkout) and push to the PR.
- When requesting changes to a PR be made, keep them within the scope of the PR.

>   - For example, while requesting small numerical adjustments on a `No C#` PR is fine, it would exceed the scope of the PR to request a new system to be added. In cases like these, it is best to close the PR and explain the changes that would need to be made.

Reviewer: ```suggestion
When leaving comments or requesting changes:
- Use **constructive** language and avoid being overly negative.
- Try not to be _overly_ vulgar or swear in a derogatory way.
- Avoid comments that personally criticize the author. 
  - It's okay to critique code and decisions, but you shouldn't insult their character.
- For people who may not be familiar with a system, try to be more thorough in your explanations.
- When possible, direct people to either the [developer wiki](https://docs.spacestation14.com/index.html), `#howdoicode` discord channel, or an alternate source for information.
- Make an effort to be available after reviewing a PR. 
  - You don't need to be constantly online, but if the author has questions, they should be able to get a response in a reasonable amount of time.
- Be formal when closing PRs. 
  - Discourage others from engaging in rude behavior and try not to upset the PR author. People can often be emotional after their PR is closed.
- For relatively minor changes, opt to simply [complete them yourself](https://cli.github.com/manual/gh_pr_checkout) and push to the PR.
- When requesting changes to a PR, keep them within the scope of the PR.
  - For example, while requesting small numerical adjustments on a `No C#` PR is fine, it would exceed the scope of the PR to request a new system to be added. In cases like these, it is best to close the PR and explain the changes that would be required.
```

- The author has obviously not read the [code conventions](../../general-development/codebase-info/conventions.md).
- The pull request's effects on the game are nil to none when thinking about the larger picture. Examples are:
  - Description changes that only make it "sound cooler" or "seem clearer" when in reality the name makes no difference.
  - Microbalancing, like small, finite adjustments of armor values, TC cost, etc. These changes should be rarely entertained.
- The pull request has not received any activity for a long time (30d+) after a requested change.
- The author has refused or is unwilling to implement the requested changes.
- The pull request has sat merge conflicted for a long time (30d+).
  - It is encouraged to request the author if they are willing to update it if you would like to start reviewing the pull request before doing so.
- The pull request is a draft even though it is not being put up for preliminary review.
- The PR has been superseded by a different PR and thus no longer needs to be merged.
- The PR author has been banned from the Wizard's Den GitHub and/or Discord or is otherwise unable to be communicated with.

> - The PR contains content which violates Wizard Den's rules and/or code of conduct.

Reviewer: ```suggestion
- The pull request's effects on the game are nil to none when thinking about the larger picture. Examples are:
  - Description changes that only make it "sound cooler" or "seem clearer" when in reality the name makes no difference.
  - Microbalancing, like small, finite adjustments of armor values, TC cost, etc. These changes should be rarely entertained.
- The pull request has not received any activity for a long time (30d+) after a requested change.
- The author has refused or is unwilling to implement the requested changes.
- The pull request has sat in a merge conflict for a long time (30d+).
  - It is encouraged to request the author if they are willing to update it if you would like to start reviewing the pull request before doing so.
- The pull request is a draft even though it is not being put up for preliminary review.
- The PR has been superseded by a different PR and thus no longer needs to be merged.
- The PR author has been banned from the Wizard's Den GitHub and/or Discord or is otherwise unable to be communicated with.
- The PR contains content that violates Wizard Den's rules and/or code of conduct.
```

This should have information about the elements that need to be addressed if a subsequent PR were to be made.

### Merging PRs
A pull request is ready to be merged if it fulfills the approval requirements associated with the PR, and all requested changes are resolved.

Next:
1. Check one final time to ensure that the changelog is correct and the breaking changes section is accurate.
2. Add the pull request to the merge queue.
3. Post the breaking changes announcement in the [Breaking Changes](https://forum.spacestation14.com/c/development/breaking-changes/70) category in `Development` on Discourse.

If the pull request gets kicked out of the merge queue due to failing tests, check if the test fail is related to the PR. If not, re-add the pull request to the merge queue.

## Reviewing Docs Pull Requests
The Docs repo has a different review policy compared to the Content repository.
Since the Docs repo can be directly pushed to by maintainers, is mostly text-based, can be easily reverted/changed, and does not impact forks, most docs pull requests can be processed by one maintainer.

### Pushing to the Docs Repo

> Maintainers have direct push access to the Docs repo and are encouraged to use it to update any information that may be outdated with up-to-date information, whether it be technical documentation, amending a design document, or just fixing up grammar.

Reviewer: ```suggestion
Maintainers have direct push access to the Docs repo and are encouraged to use it to update any outdated information, whether it's technical documentation, amending a design document, or just fixing grammar.
```

### PR Reviews
PRs that update instructional or reference documentation (including but not limited to setup guides, style guides, system documentation) require only a single approval is needed before it can be merged.
Likewise, only a single maintainer needs to express concern to close it.

> Note that this does not apply to changes to internal procedure or other modifications which require voting or group deliberation according to relevant policy.

Reviewer: ```suggestion
Note that this does not apply to changes to internal procedure or other modifications that require voting or group deliberation according to relevant policy.
```

#### Design Documents

> Major PRs that introduce or change a design document should seek either two maintainer approvals or more.

Reviewer: ```suggestion
Major PRs that introduce or change a design document should require at least two maintainer approvals.
```

> If the addition is large enough, a maintainer can call a vote for the design document, however this should only be done when the scope of the pull request warrants it.

Reviewer: ```suggestion
If the addition is large enough, a maintainer can call a vote on the design document; however, this should be done only when the scope of the pull request warrants it.
```

# PR Review Procedure

## Abstract
This documents outlines the Maintainer procedure for reviewing and merging PRs to the [Space Station 14 repository](https://github.com/space-wizards/space-station-14) `master` branch (otherwise known as the Content repository).

Pull requests targeting the `staging` or `stable` branch are subject to the [Hotfix Procedure](hotfix-procedure.md).

Any portion of this procedure may be waived and/or modified with written permission (_via Discord or Github_) from a Lead Maintainer, Project Manager or Wizard.

```admonish note
Note that these procedures are ultimately flexible, however Maintainers that deviate from this policy are expected to provide concrete and valid reasoning when doing so.
Getting one or more other maintainer(s) to agree with your deviation is advisable.

> Note that all pull requests that are scheduled for release to `stable` are reviewed during _Stable Review_ in maintainer meetings.

Reviewer 2: This is only the case for Feature PRs or sometimes particularly notable refactors, and of those, nowadays we generally only review the ones that had comments on them (although recently, we had deviated from that sometimes, when we had a quiet week so it took less time to go through them all)

If the intent is to communicate that they can still be changed or reverted, as noted above, then it could just say `are reviewed before or during _Stable Review_` which is a small change, but we probably want the policy to accurately say what we are doing

```

## Reviewing Content Pull Requests

### Reviewing the PR

> All pull requests must adhere to the [code conventions](../../general-development/codebase-info/conventions.md).

Reviewer 3: ```suggestion
If PR files do not adhere to the [code conventions](../../general-development/codebase-info/conventions.md) or the PR Template is not fully filled out by the author, you may close the PR.
```

All parts of the Pull Request Template must be filled out by the author.

If the author has obviously not read the conventions or PR guidelines, or the pull request template has not been filled out to a satisfactory level, close the PR.

> All PRs must be triaged, as per the [triage procedure](triage-procedure.md) guidelines.

Reviewer 3: ```suggestion
When reviewing an unlabelled PR, you must triage the PR or request triage. PRs must be triaged according to [triage procedure](triage-procedure.md) guidelines.
```

> You should triage the pull request when reviewing it for the first time.

Reviewer 3: ```suggestion
```

When leaving comments or requesting changes:
- Use **constructive** language and avoid being overly negative.
- Try not to be _overly_ vulgar or swear in a derogatory way.

> - Avoid comments that personally criticize the author. 

Reviewer 3: ```suggestion
- Avoid insulting or criticizing the author. (You are reviewing the PR, not their character.)
```

  - It's okay to critique code and decisions, but you shouldn't insult their character.
- For people who may not be familiar with a system, try to be more thorough in your explanations.

> - When possible, direct people to either the [developer wiki](https://docs.spacestation14.com/index.html), `#howdoicode` discord channel, or alternate source for information.

Reviewer 3: ```suggestion
- Direct people to the [developer wiki](https://docs.spacestation14.com/index.html), `#howdoicode` Discord channel, or SS14 Discourse 'Development' section.
```

- Make an effort to be available after reviewing a PR. 
  - You don't need to be constantly online, but if the author has questions, they should be able to get a response in a reasonable amount of time.

> - Be formal when closing PRs. 

Reviewer 3: ```suggestion
```
Formality is nice but kinda unnecessary.

  - Discourage others from engaging in rude behavior and try not to upset the PR author. People can often be emotional after their PR is closed.
- For relatively minor changes, opt to simply [complete them yourself](https://cli.github.com/manual/gh_pr_checkout) and push to the PR.

> - When requesting changes to a PR be made, keep them within the scope of the PR.

Reviewer 3: ```suggestion
- Keep requested changes within the scope of the PR, unless you have requested the author to split their PR.
```

>   - For example, while requesting small numerical adjustments on a `No C#` PR is fine, it would exceed the scope of the PR to request a new system to be added. In cases like these, it is best to close the PR and explain the changes that would need to be made.

Reviewer 3: ```suggestion
  - For example, while requesting small numerical adjustments on a `No C#` PR is fine, you would exceed the scope of the PR to request a new system to be added. In such cases, it is best to close the PR and explain the changes that would need to be made.
  
You may [complete minor changes yourself](https://cli.github.com/manual/gh_pr_checkout) and push a new commit to the PR.
```

### Approving PRs

> Pull requests must pass code review in order to be merged.

Reviewer 3: ```suggestion
PRs must pass code review to be merged.
```

The number of reviews required are determined by the type of pull request being reviewed.

```admonish note
Pull requests may be reviewed, approved, and merged by a single maintainer in specific circumstances. When a maintainer does so, the maintainer is expected to provide concrete and valid reasoning for doing so. For example, a maintainer that is known for reviewing and making prediction PRs can probably process a prediction PR by themselves.
```

If the pull request does not line up with any category listed here, defer to two approvals for a pull request.

A pull request created by a Maintainer has one approval by default.
However, Maintainers cannot self-approve pull requests that would only require one approval to merge.

> For example, a maintainer cannot create a PR for a small bugfix and then instantly merge it after - it must be reviewed by another maintainer.

Reviewer 3: ```suggestion

A Maintainer PR has one approval by default. However, all PRs require a second review from another Maintainer. (Maintainers cannot self-approve and self-merge.)
```

    - These should usually come with a custom BenchmarkDotNet benchmark to prove that the PR has a meaningful effect on performance.
  - Player-facing changes large in impact.
    - Major features, new content.

#### Exceptions
There are some exceptions to the one-approval or two-approval system. They are listed below.
1. **Mapping changes must be approved by a Mapping Lead, and only require one approval if the PR contains only mapping changes.**
   1. This still applies even if the change is technical in origin (like fixing a datafield spamming/inflating the size of YAML).
   2. Mapping Leads can self-approve their own mapping changes.
2. **Sprite changes must be approved by an Art Lead, and only require one approval if the PR contains only sprite changes.**
   1. Art Lead approval is not required in the case of a sprite being minorly fixed up (for example, translation/rotation, removing stray pixels, etc.).
   2. Art Leads cannot self-approve sprite changes.
3. **Changes targeting a [Maintainer Workgroups](maintainer-workgroup-policy.md)'s game area must be approved by a member from the workgroup.**
   1. Workgroup members can self-approve changes targeting their own workgroup.

> 4. **Rule or server configuration changes require the approval, either written or on GitHub, of the Head Game Admin team.**

Reviewer 4: ```suggestion
4. **Rule or server configuration changes require the approval, either written or on GitHub, of the Head Game Admin or infrastructure team.**
```

There are some configs game admins don't really need to approve but infra (everyone with ssh access to the servers) may want to make. May as well put it here.

### Closing PRs

> Pull requests can be closed in two ways:

Reviewer 3: ```suggestion
PRs can be closed in two ways:
```

- A maintainer expresses **concern** using the `S: Concern` label on GitHub. The amount of maintainers disapproving of the change required to close the PR is the same amount that would be required to merge the pull request.
  - For example, PRs that are classified as small content additions can be closed by one maintainer.
- It meets any one of the requirements to close listed in the below section.

#### Other Reasons for Closing

> - As previously mentioned, the pull request body is not filled out to an acceptable level given the changes.

Reviewer 3: ```suggestion
```
Redundant header and text

> - The author has obviously not read the [code conventions](../../general-development/codebase-info/conventions.md).

Reviewer 3: ```suggestion
```
Redundant

> - The author has obviously not read the [code conventions](../../general-development/codebase-info/conventions.md).

Reviewer 3: ```suggestion
- The PR template is not filled out.
- The PR files do not follow [code conventions](../../general-development/codebase-info/conventions.md).
```
Redundant header if we go from 'two ways' to 'for any of the following reasons'

> - The pull request's effects on the game are nil to none when thinking about the larger picture. Examples are:

Reviewer 3: ```suggestion
- Gameplay effects are inconsequential in the bigger picture. Examples include:
```

  - Description changes that only make it "sound cooler" or "seem clearer" when in reality the name makes no difference.
  - Microbalancing, like small, finite adjustments of armor values, TC cost, etc. These changes should be rarely entertained.
- The pull request has not received any activity for a long time (30d+) after a requested change.
- The author has refused or is unwilling to implement the requested changes.
- The pull request has sat merge conflicted for a long time (30d+).

>   - It is encouraged to request the author if they are willing to update it if you would like to start reviewing the pull request before doing so.

Reviewer 3: ```suggestion
```
Section acts mainly to describe methods of closing PRs, so this point seems moot.

This invalidates all disagreements/`S: Concern` tags and allows the pull request to be merged as long as it passes code review.

If the discussion has a negative conclusion, the closure message for the PR should include a brief summary of the discussion.
This should have information about the elements that need to be addressed if a subsequent PR were to be made.

### Tagging a PR as Experimental
An Experimental PR is any PR marked with the `Intent: Experimental` label.
This label is assigned at a Maintainer's discretion and indicates the PR contains features or changes that are:
1. Expected to get a lot of feedback from players.
2. The Maintainer team is seeking a lot of feedback on from playtesting.

While all PRs benefit from feedback and may be changed at a later time, Experimental PRs are intended to indicate to the community that feedback is especially appreciated.

Adding the label to a PR will create a forum thread with the PR's name and number in the Feedback category of the forum. Once merged, any changelog entry for the PR will have an additional :test_tube: symbol to indicate its Experimental status.

> It is strongly recommended that a `FeedbackPopup` prototype for the thread is added to `Resources/Prototypes/FeedbackPopup/feedbackpopups.yml`.

Reviewer 3: ```suggestion
The maintainer that merges the Experimental PR must add a `FeedbackPopup` prototype with Discourse thread URL to `Resources/Prototypes/FeedbackPopup/feedbackpopups.yml`.
```
I assume whoever merges the Experimental PR has to create the FeedbackPopup prototype.

### Merging PRs
A pull request is ready to be merged if it fulfills the approval requirements associated with the PR, and all requested changes are resolved.

Next:
1. Check one final time to ensure that the changelog is correct and the breaking changes section is accurate.
2. Add the pull request to the merge queue.

> 3. Post the breaking changes announcement in the [Breaking Changes](https://forum.spacestation14.com/c/development/breaking-changes/70) category in `Development` on Discourse.

Reviewer 3: ```suggestion
1. Check one final time to ensure that the changelog is correct and the breaking changes section is accurate.
2. Add the pull request to the merge queue.
3. Check if the PR test failure is related. If not, re-add the pull request to the merge queue.
4. Post the breaking changes announcement in the [Breaking Changes](https://forum.spacestation14.com/c/development/breaking-changes/70) category in `Development` on Discourse.
```

> If the pull request gets kicked out of the merge queue due to failing tests, check if the test fail is related to the PR. If not, re-add the pull request to the merge queue.

Reviewer 2: This would be a good place to mention that we have a megaissue for tracking known Heisentests, which hopefully makes each one get fixed faster
https://github.com/space-wizards/space-station-14/issues/41081
